source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -0,0 +1,79 @@
+#include "source/extensions/filters/http/header_mutation/header_mutation.h"
+
+#include <cstdint>
+#include <memory>
+
+#include "source/common/config/utility.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/common/http/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HeaderMutation {
+
+void Mutations::mutateRequestHeaders(Http::RequestHeaderMap& request_headers,
+                                     const StreamInfo::StreamInfo& stream_info) const {
+  request_mutations_.evaluateHeaders(request_headers, request_headers,
+                                     *Http::StaticEmptyHeaders::get().response_headers,
+                                     stream_info);
+}
+
+void Mutations::mutateResponseHeaders(const Http::RequestHeaderMap& request_headers,
+                                      Http::ResponseHeaderMap& response_headers,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  response_mutations_.evaluateHeaders(response_headers, request_headers, response_headers,
+                                      stream_info);
+}
+
+PerRouteHeaderMutation::PerRouteHeaderMutation(const PerRouteProtoConfig& config)
+    : mutations_(config.mutations()) {}
+
+HeaderMutationConfig::HeaderMutationConfig(const ProtoConfig& config)
+    : mutations_(config.mutations()) {}
+
+Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  config_->mutations().mutateRequestHeaders(headers, decoder_callbacks_->streamInfo());
+
+  // Only the most specific route config is used.
+  // TODO(wbpcode): It's possible to traverse all the route configs to merge the header mutations
+  // in the future.
+  route_config_ =
+      Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteHeaderMutation>(decoder_callbacks_);
+
+  if (route_config_ != nullptr) {
+    route_config_->mutations().mutateRequestHeaders(headers, decoder_callbacks_->streamInfo());
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus HeaderMutation::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
+  // The request headers will be set to the downstream stream info before the filter chain is
+  // started. And in the current implementation, the upstream filter chain will reuse the downstream
+  // stream info. So the getRequestHeaders() will never return nullptr no matter the filter is
+  // is used as a downstream or upstream filter.
+  ASSERT(encoder_callbacks_->streamInfo().getRequestHeaders() != nullptr);
```

Comment:
Can we relax this assert here? Seems it is breaking [filter_fuzz_test](https://github.com/envoyproxy/envoy/blob/30cfa5c4619bae49125fff35aacf2f62e81b4410/test/extensions/filters/http/common/fuzz/BUILD#L66) with this corpus:
```
config {
  name: "envoy.filters.http.header_mutation"
  typed_config {
    type_url: "type.googleapis.com/envoy.extensions.filters.http.header_mutation.v3.HeaderMutation"
  }
}
```
Or we may need to change the fuzz test infra? Can you take a look? Thanks!

Reply:
I think we need a fix here. I just get some new knowledge about the filter chain. It's possible to get a null request headers in the encoding phase.

Reply:
Thanks!